### ISSUE TITLE
WGTokenInvalidException still occurs after session refresh sometimes, for some reason

### ISSUE THREAD

**Issue by dyc3** [dyc3/steamguard-cli#28]
Related to #11

**Comment by melvyn2** [comment]
here it is:
```
Action: accept-all
maFiles path: path/to/maFiles
Opening manifest...
1 accounts loaded
Reading accounts from manifest...
Accepting Confirmations on user
Refeshing Session...
Session refreshed
Saving account user to number.maFile...
Fetching Confirmations...

Unhandled Exception:
System.AggregateException: One or more errors occurred. ---> SteamAuth.SteamGuardAccount+WGTokenInvalidException: Exception of type 'SteamAuth.SteamGuardAccount+WGTokenInvalidException' was thrown.
  at SteamAuth.SteamGuardAccount+<FetchConfirmationsAsync>d__57.MoveNext () [0x0015f] in <2b9176aa3dc7472eb5981f87737f7e31>:0 
   --- End of inner exception stack trace ---
  at System.Threading.Tasks.Task.ThrowIfExceptional (System.Boolean includeTaskCanceledExceptions) [0x00011] in <6e9b92f0d119424382ef180639777acb>:0 
  at System.Threading.Tasks.Task.Wait (System.Int32 millisecondsTimeout, System.Threading.CancellationToken cancellationToken) [0x00043] in <6e9b92f0d119424382ef180639777acb>:0 
  at System.Threading.Tasks.Task.Wait () [0x00000] in <6e9b92f0d119424382ef180639777acb>:0 
  at SteamGuard.Program.AcceptAllTrades (System.String user, System.String passkey) [0x00156] in <c9b4adbb23d14efcafaf3c06134ccbfc>:0 
  at SteamGuard.Program.Main (System.String[] args) [0x00375] in <c9b4adbb23d14efcafaf3c06134ccbfc>:0 
---> (Inner Exception #0) SteamAuth.SteamGuardAccount+WGTokenInvalidException: Exception of type 'SteamAuth.SteamGuardAccount+WGTokenInvalidException' was thrown.
  at SteamAuth.SteamGuardAccount+<FetchConfirmationsAsync>d__57.MoveNext () [0x0015f] in <2b9176aa3dc7472eb5981f87737f7e31>:0 <---

[ERROR] FATAL UNHANDLED EXCEPTION: System.AggregateException: One or more errors occurred. ---> SteamAuth.SteamGuardAccount+WGTokenInvalidException: Exception of type 'SteamAuth.SteamGuardAccount+WGTokenInvalidException' was thrown.
  at SteamAuth.SteamGuardAccount+<FetchConfirmationsAsync>d__57.MoveNext () [0x0015f] in <2b9176aa3dc7472eb5981f87737f7e31>:0 
   --- End of inner exception stack trace ---
  at System.Threading.Tasks.Task.ThrowIfExceptional (System.Boolean includeTaskCanceledExceptions) [0x00011] in <6e9b92f0d119424382ef180639777acb>:0 
  at System.Threading.Tasks.Task.Wait (System.Int32 millisecondsTimeout, System.Threading.CancellationToken cancellationToken) [0x00043] in <6e9b92f0d119424382ef180639777acb>:0 
  at System.Threading.Tasks.Task.Wait () [0x00000] in <6e9b92f0d119424382ef180639777acb>:0 
  at SteamGuard.Program.AcceptAllTrades (System.String user, System.String passkey) [0x00156] in <c9b4adbb23d14efcafaf3c06134ccbfc>:0 
  at SteamGuard.Program.Main (System.String[] args) [0x00375] in <c9b4adbb23d14efcafaf3c06134ccbfc>:0 
---> (Inner Exception #0) SteamAuth.SteamGuardAccount+WGTokenInvalidException: Exception of type 'SteamAuth.SteamGuardAccount+WGTokenInvalidException' was thrown.
  at SteamAuth.SteamGuardAccount+<FetchConfirmationsAsync>d__57.MoveNext () [0x0015f] in <2b9176aa3dc7472eb5981f87737f7e31>:0 <---

```

**Comment by dyc3** [comment]
To be honest, I'm not quite sure whats causing this. I want to know if your updated session is being saved to disk. Run these commands and post their output:
* `md5sum path/to/number.maFile`
* `steamguard-cli -v trade`
* `md5sum path/to/number.maFile`

**Comment by melvyn2** [comment]
Ok:

```da27b04dfc045f462aeb07d312ba3336  path/to/maFiles/number.maFile
Action: accept-all
maFiles path: path/to/maFiles/
Opening manifest...
1 accounts loaded
Reading accounts from manifest...
Accepting Confirmations on user
Refeshing Session...
Session refreshed
Saving account user to number.maFile...
Fetching Confirmations...

Unhandled Exception:
System.AggregateException: One or more errors occurred. ---> SteamAuth.SteamGuardAccount+WGTokenInvalidException: Exception of type 'SteamAuth.SteamGuardAccount+WGTokenInvalidException' was thrown.
  at SteamAuth.SteamGuardAccount+<FetchConfirmationsAsync>d__57.MoveNext () [0x0015f] in <2b9176aa3dc7472eb5981f87737f7e31>:0 
   --- End of inner exception stack trace ---
  at System.Threading.Tasks.Task.ThrowIfExceptional (System.Boolean includeTaskCanceledExceptions) [0x00011] in <6e9b92f0d119424382ef180639777acb>:0 
  at System.Threading.Tasks.Task.Wait (System.Int32 millisecondsTimeout, System.Threading.CancellationToken cancellationToken) [0x00043] in <6e9b92f0d119424382ef180639777acb>:0 
  at System.Threading.Tasks.Task.Wait () [0x00000] in <6e9b92f0d119424382ef180639777acb>:0 
  at SteamGuard.Program.AcceptAllTrades (System.String user, System.String passkey) [0x00156] in <c9b4adbb23d14efcafaf3c06134ccbfc>:0 
  at SteamGuard.Program.Main (System.String[] args) [0x00375] in <c9b4adbb23d14efcafaf3c06134ccbfc>:0 
---> (Inner Exception #0) SteamAuth.SteamGuardAccount+WGTokenInvalidException: Exception of type 'SteamAuth.SteamGuardAccount+WGTokenInvalidException' was thrown.
  at SteamAuth.SteamGuardAccount+<FetchConfirmationsAsync>d__57.MoveNext () [0x0015f] in <2b9176aa3dc7472eb5981f87737f7e31>:0 <---

[ERROR] FATAL UNHANDLED EXCEPTION: System.AggregateException: One or more errors occurred. ---> SteamAuth.SteamGuardAccount+WGTokenInvalidException: Exception of type 'SteamAuth.SteamGuardAccount+WGTokenInvalidException' was thrown.
  at SteamAuth.SteamGuardAccount+<FetchConfirmationsAsync>d__57.MoveNext () [0x0015f] in <2b9176aa3dc7472eb5981f87737f7e31>:0 
   --- End of inner exception stack trace ---
  at System.Threading.Tasks.Task.ThrowIfExceptional (System.Boolean includeTaskCanceledExceptions) [0x00011] in <6e9b92f0d119424382ef180639777acb>:0 
  at System.Threading.Tasks.Task.Wait (System.Int32 millisecondsTimeout, System.Threading.CancellationToken cancellationToken) [0x00043] in <6e9b92f0d119424382ef180639777acb>:0 
  at System.Threading.Tasks.Task.Wait () [0x00000] in <6e9b92f0d119424382ef180639777acb>:0 
  at SteamGuard.Program.AcceptAllTrades (System.String user, System.String passkey) [0x00156] in <c9b4adbb23d14efcafaf3c06134ccbfc>:0 
  at SteamGuard.Program.Main (System.String[] args) [0x00375] in <c9b4adbb23d14efcafaf3c06134ccbfc>:0 
---> (Inner Exception #0) SteamAuth.SteamGuardAccount+WGTokenInvalidException: Exception of type 'SteamAuth.SteamGuardAccount+WGTokenInvalidException' was thrown.
  at SteamAuth.SteamGuardAccount+<FetchConfirmationsAsync>d__57.MoveNext () [0x0015f] in <2b9176aa3dc7472eb5981f87737f7e31>:0 <---

868f350b5501a7e12aabf2e9fa9de6ce  path/to/maFiles/number.maFile```

**Comment by dyc3** [comment]
Does your `maFile` work with the latest version of SDA?

**Comment by melvyn2** [comment]
SDA currently dosent work on my mac: it just freezes on startup.

**Comment by amdev0** [comment]
Apparently it's an issue with SteamAuth which relied on extracting text from HTML via regex to process confirmations and Steam decided to change their descriptions.
Updating the SteamAuth submodule with the latests commits fixes this with very few changes needed.

**Comment by melvyn2** [comment]
Ok, I'll try that

**Comment by dyc3** [comment]
Sorry for taking so long on this one, I've been on spring break and I was spending time with family. Thanks @amdev0 for the info, I'll do some testing on my end, (and I would like to hear from @melvyn2) before closing this.

**Comment by melvyn2** [comment]
Actually, when I update the submodule (`cd SteamAuth; git pull origin master`), I get these errors (full log [here](https://pastebin.com/mdKJeQ2w)):
```
Program.cs(664,73): error CS1061: Type `SteamAuth.Confirmation' does not contain a definition for `Description' and no extension method `Description' of type `SteamAuth.Confirmation' could be found. Are you missing an assembly reference?
/tmp/steamguard-cli/build/SteamAuth.dll (Location of the symbol related to previous error)
Program.cs(707,67): error CS1061: Type `SteamAuth.Confirmation' does not contain a definition for `Description' and no extension method `Description' of type `SteamAuth.Confirmation' could be found. Are you missing an assembly reference?
/tmp/steamguard-cli/build/SteamAuth.dll (Location of the symbol related to previous error)
Program.cs(711,63): error CS1061: Type `SteamAuth.Confirmation' does not contain a definition for `Description' and no extension method `Description' of type `SteamAuth.Confirmation' could be found. Are you missing an assembly reference?
/tmp/steamguard-cli/build/SteamAuth.dll (Location of the symbol related to previous error)
Program.cs(715,65): error CS1061: Type `SteamAuth.Confirmation' does not contain a definition for `Description' and no extension method `Description' of type `SteamAuth.Confirmation' could be found. Are you missing an assembly reference?
/tmp/steamguard-cli/build/SteamAuth.dll (Location of the symbol related to previous error)
```

**Comment by melvyn2** [comment]
Without any idea with what I was doing, I removed the offending lines, because I couldn't find any way to get the description of the trade in the latest steamauth, and it worked! Thank you @amdev0 !

**Comment by amdev0** [comment]
Yes, they changed the Confirmation class so it no longer contains the description of a trade and that's what I meant with "very few changes needed". I don't know if there's another simple way of getting trade descriptions currently.
Anyway, glad it worked :thumbsup:

**Comment by dyc3** [comment]
Thank you, both for helping out, I really appreciate it. Before I release a new version, I'm hoping to retain some usability by finding a replacement to the `Confirmation.Description` property that was removed from SteamAuth.

**Comment by melvyn2** [comment]
I've looked through their code, and I haven't found anything... I actually haven't seen the lines that rely on description come in to play, and saw no difference when I removed them. _My **opinion**_ is that a release or pre-release that fixes a sort-of major problem is worth having, even if it removes a little bit of information. But that's just my opinion ;)

**Comment by dyc3** [comment]
That makes sense. If I can't find an easy solution within a few hours, then I'll release a new version with the fix.

**Comment by amdev0** [comment]
I agree this issue is a pretty big deal because it makes trading impossible and should be fixed as soon as possible. For the time being `Confirmation.ConfType` along with an ID like `Confirmation.Creator` could be used instead, just a suggestion.

**Comment by melvyn2** [comment]
Good to hear that 👍

**Comment by dyc3** [comment]
This should be fixed in [v0.3.5.0](https://github.com/dyc3/steamguard-cli/releases/tag/v0.3.5.0)
